File: scripts/m49.mjs
import { mkdirSync, writeFileSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const OUTPUT_DIR = join(__dirname, "../public/api");
mkdirSync(OUTPUT_DIR, { recursive: true });
const OUTPUT = join(OUTPUT_DIR, "m49.csv");

const URL = "https://unstats.un.org/unsd/methodology/m49/overview/";
const TABLE_ID = "downloadTableEN";

console.log(`Fetching ${URL} ...`);
const res = await fetch(URL, {
  headers: { "User-Agent": "Mozilla/5.0" },
});
if (!res.ok) throw new Error(`HTTP ${res.status} ${res.statusText}`);
const html = await res.text();

// Extract the target table's HTML
const tableRe = new RegExp(`<table[^>]*id\\s*=\\s*"${TABLE_ID}"[^>]*>([\\s\\S]*?)</table>`, "i");
const tableMatch = html.match(tableRe);
if (!tableMatch) throw new Error(`Table #${TABLE_ID} not found in page`);
const tableHtml = tableMatch[0];

// Parse rows and cells
const rows = [];
for (const rowMatch of tableHtml.matchAll(/<tr[\s>]([\s\S]*?)<\/tr>/gi)) {
  const cells = [];
  for (const cellMatch of rowMatch[1].matchAll(/<t[dh][\s>]([\s\S]*?)<\/t[dh]>/gi)) {
    // Strip inner tags and decode basic HTML entities
    const text = cellMatch[1]
      .replace(/<[^>]+>/g, "")
      .replace(/&amp;/g, "&")
      .replace(/&lt;/g, "<")
      .replace(/&gt;/g, ">")
      .replace(/&nbsp;/g, " ")
      .replace(/&#(\d+);/g, (_, code) => String.fromCharCode(Number(code)))
      .trim();
    cells.push(text);
  }
  if (cells.length) rows.push(cells);
}

if (rows.length < 2) throw new Error(`Only ${rows.length} rows parsed — check the HTML structure`);

// Convert header row to snake_case
rows[0] = rows[0].map((h) =>
  h
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_|_$/g, ""),
);

// Serialize to CSV
function csvRow(/** @type {string[]} */ cells) {
  return cells
    .map((/** @type {string} */ c) =>
      c.includes(",") || c.includes('"') || c.includes("\n") ? `"${c.replace(/"/g, '""')}"` : c,
    )
    .join(",");
}

const csv = rows.map(csvRow).join("\n") + "\n";
writeFileSync(OUTPUT, csv, "utf8");
console.log(`Wrote ${rows.length - 1} rows → ${OUTPUT}`);

File: scripts/m49.py
# /// script
# requires-python = ">=3.12"
# dependencies = ["requests"]
# ///

import csv
import re
import sys
from pathlib import Path

import requests

OUTPUT_DIR = Path(__file__).parent.parent / "public" / "api"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT = OUTPUT_DIR / "m49.csv"

URL = "https://unstats.un.org/unsd/methodology/m49/overview/"
TABLE_ID = "downloadTableEN"

print(f"Fetching {URL} ...")
res = requests.get(URL, headers={"User-Agent": "Mozilla/5.0"}, timeout=30)
res.raise_for_status()
html = res.text

# Extract the target table's HTML
table_re = re.compile(
    rf'<table[^>]*id\s*=\s*"{TABLE_ID}"[^>]*>([\s\S]*?)</table>', re.IGNORECASE
)
table_match = table_re.search(html)
if not table_match:
    print(f"Error: Table #{TABLE_ID} not found in page", file=sys.stderr)
    sys.exit(1)
table_html = table_match.group(0)


def decode_cell(raw: str) -> str:
    text = re.sub(r"<[^>]+>", "", raw)
    text = text.replace("&amp;", "&")
    text = text.replace("&lt;", "<")
    text = text.replace("&gt;", ">")
    text = text.replace("&nbsp;", " ")
    text = re.sub(r"&#(\d+);", lambda m: chr(int(m.group(1))), text)
    return text.strip()


rows = []
for row_match in re.finditer(r"<tr[\s>]([\s\S]*?)</tr>", table_html, re.IGNORECASE):
    cells = [
        decode_cell(cell_match.group(1))
        for cell_match in re.finditer(
            r"<t[dh][\s>]([\s\S]*?)</t[dh]>", row_match.group(1), re.IGNORECASE
        )
    ]
    if cells:
        rows.append(cells)

if len(rows) < 2:
    print(f"Error: Only {len(rows)} rows parsed — check the HTML structure", file=sys.stderr)
    sys.exit(1)

# Convert header row to snake_case
rows[0] = [
    re.sub(r"^_|_$", "", re.sub(r"[^a-z0-9]+", "_", h.lower())) for h in rows[0]
]

with open(OUTPUT, "w", newline="", encoding="utf-8") as f:
    writer = csv.writer(f)
    writer.writerows(rows)

print(f"Wrote {len(rows) - 1} rows → {OUTPUT}")
